bigtable/bigtable-proxy/src/test/java/com/google/cloud/bigtable/examples/proxy/commands/ServeTest.java

```diff
@@ -0,0 +1,426 @@
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.stub.StreamObserver;
+import io.grpc.testing.GrpcCleanupRule;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.util.UUID;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class ServeTest {
+  private final String targetServerName = UUID.randomUUID().toString();
+
+  @Rule
+  public final GrpcCleanupRule grpcCleanup = new GrpcCleanupRule().setTimeout(1, TimeUnit.MINUTES);
+
+  // Fake targets
+  private MetadataInterceptor metadataInterceptor;
+  private FakeDataService dataService;
+  private FakeInstanceAdminService instanceAdminService;
+  private FakeTableAdminService tableAdminService;
+  private OperationService operationService;
+  private ManagedChannel fakeServiceChannel;
+
+  // Proxy
+  private Serve serve;
+  private ManagedChannel proxyChannel;
+
+  @Before
+  public void setUp() throws IOException {
+    // Create the fake target
+    metadataInterceptor = new MetadataInterceptor();
+    dataService = new FakeDataService();
+    instanceAdminService = new FakeInstanceAdminService();
+    tableAdminService = new FakeTableAdminService();
+    operationService = new OperationService();
+
+    grpcCleanup.register(
+        InProcessServerBuilder.forName(targetServerName)
+            .intercept(metadataInterceptor)
+            .addService(dataService)
+            .addService(instanceAdminService)
+            .addService(tableAdminService)
+            .addService(operationService)
+            .build()
+            .start());
+
+    fakeServiceChannel =
+        grpcCleanup.register(
+            InProcessChannelBuilder.forName(targetServerName).usePlaintext().build());
+
+    // Create the proxy
+    serve = createAndStartCommand(fakeServiceChannel);
+
+    proxyChannel =
+        grpcCleanup.register(
+            ManagedChannelBuilder.forAddress("localhost", serve.listenPort).usePlaintext().build());
+  }
+
+  @After
+  public void tearDown() throws InterruptedException {
+    if (serve != null) {
+      serve.cleanup();
+    }
+  }
+
+  @Test
+  public void testDataRpcOk() throws InterruptedException, ExecutionException, TimeoutException {
+    BigtableFutureStub proxyStub = BigtableGrpc.newFutureStub(proxyChannel);
+
```

Comment:
Should we also check readRows for sever streaming?

Reply:
I dont think so...the proxy is completely oblivious to unary vs server streaming